driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java

```diff
@@ -247,46 +246,43 @@ private boolean handleReplicaSetMemberChanged(final ServerDescription newDescrip
-            }
+            LOGGER.info(format("Canonical address %s does not match server address.  Removing %s from client view of cluster",
+                    newDescription.getCanonicalAddress(), newDescription.getAddress()));
             removeServer(newDescription.getAddress());
             return true;
         }
 
         if (newDescription.isPrimary()) {
             ObjectId electionId = newDescription.getElectionId();
             Integer setVersion = newDescription.getSetVersion();
-            if (setVersion != null && electionId != null) {
-                if (isStalePrimary(newDescription)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info(format("Invalidating potential primary %s whose (set version, election id) tuple of (%d, %s) "
-                                + "is less than one already seen of (%d, %s)",
-                                newDescription.getAddress(),
-                                setVersion, electionId,
-                                maxSetVersion, maxElectionId));
-                    }
-                    addressToServerTupleMap.get(newDescription.getAddress()).server.resetToConnecting();
+            if (newDescription.getMaxWireVersion() >= SIX_DOT_ZERO_WIRE_VERSION) {
+                if (nullSafeCompareTo(electionId, maxElectionId) > 0
+                        || (nullSafeCompareTo(electionId, maxElectionId) == 0 && nullSafeCompareTo(setVersion, maxSetVersion) >= 0)) {
+                    LOGGER.info(format("Setting max election id to %s and max set version to %d from replica set primary %s",
+                            newDescription.getElectionId(), newDescription.getSetVersion(), newDescription.getAddress()));
+                    maxElectionId = electionId;
+                    maxSetVersion = setVersion;
+                } else {
+                    invalidatePotentialPrimary(newDescription);
                     return false;
                 }
-
-                if (!electionId.equals(maxElectionId)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info(format("Setting max election id to %s from replica set primary %s", electionId,
-                                newDescription.getAddress()));
+            } else {
+                if (setVersion != null && electionId != null) {
+                    if (nullSafeCompareTo(maxSetVersion, setVersion) <= 0
+                            && (nullSafeCompareTo(maxSetVersion, setVersion) != 0 || nullSafeCompareTo(maxElectionId, electionId) <= 0)) {
+                        LOGGER.info(format("Setting max election id to %s from replica set primary %s",
+                                newDescription.getElectionId(), newDescription.getAddress()));
+                        maxElectionId = electionId;
+                    } else {
+                        invalidatePotentialPrimary(newDescription);
+                        return false;
                     }
```

Comment:
It appears to me, that the following code fragment is easier to digest and to compare with the original code, than the one highlighted by this comment:

```java
// current
if (nullSafeCompareTo(maxSetVersion, setVersion) <= 0
        && (nullSafeCompareTo(maxSetVersion, setVersion) != 0 || nullSafeCompareTo(maxElectionId, electionId) <= 0)) {
    LOGGER.info(format("Setting max election id to %s from replica set primary %s",
            newDescription.getElectionId(), newDescription.getAddress()));
    maxElectionId = electionId;
} else {
    invalidatePotentialPrimary(newDescription);
    return false;
}
```
```java
// proposed equivalent
if (setVersion != null && electionId != null) {
    if (nullSafeCompareTo(maxSetVersion, setVersion) > 0
            || (nullSafeCompareTo(maxSetVersion, setVersion) == 0 && nullSafeCompareTo(maxElectionId, electionId) > 0)) {
        invalidatePotentialPrimary(newDescription);
        return false;
    }
    LOGGER.info(format("Setting max election id to %s from replica set primary %s",
            newDescription.getElectionId(), newDescription.getAddress()));
    maxElectionId = electionId;
}
```

The proposed fragment also makes it clear that the new code has a different logging behavior: previously `"Setting max election id to %s from replica set primary %s"` was only logged if `!electionId.equals(maxElectionId)`, but now it is logged regardless of this condition.

The proposed fragment is also much simpler to compare with the new logic used if `newDescription.getMaxWireVersion() >= SIX_DOT_ZERO_WIRE_VERSION`, and see that the tuple elements are compared in a different order.

Reply:
See my comment [here](https://github.com/mongodb/mongo-java-driver/pull/1010#discussion_r988344581) for the justification of the current logic.  If it's not compelling, I'll switch it around again.

Reply:
In my opinion the "unreversed" code is much simpler to read (I was unable to understand the condition until I reversed it manually), and is more alike with the 6.0+ case above. Maybe it's just me.

Reply:
Sorry, I missed your comment on the changed logging behavior.  That change is unintentional, on both conditional branches, and needs to be reverted.  I did that, and also extracted some methods that hopefully clarify the intention and also why I organized the code this way at a high level.

Reply:
I still wasn't happy with it, so I tried again.  I think the new code flow is much simpler, though it diverges even more from the spec pseudocode.  CC @nbbeeken 

I also simplified the logging to just a single statement, instead of three.  In practice, you'll only get a new electionId or setVersion with a new primary, so a single statement is justified IMO (the logging is also not yet part of the spec, so we can do what we want here).

Reply:
> I think the new code flow is much simpler, though it diverges even more from the spec pseudocode.

I think this simply means that it may be a good idea to refactor the the spec pseudocode.